### ISSUE TITLE
Crash getting volumeClaimTemplates creation timestamp

### ISSUE THREAD

**Issue by roderik** [pulumi/pulumi-kubernetes#2359]
### What happened?

I have a deployed stack, was trying to refresh/up, pulumi crashed

```
Previewing update (settlemint/public-nodes-k8s)

View in Browser (Ctrl+O): https://app.pulumi.com/settlemint/bpaas/public-nodes-k8s/previews/c93ffb93-eea1-43c6-b086-0e6efd1d80f2

     Type                                                                 Name                                     Plan        Info
     pulumi:pulumi:Stack                                                  bpaas-public-nodes-k8s                               44 messages
 ~   ├─ pulumi:providers:kubernetes                                       k8s                                      update      [diff: ~version]
     ├─ kubernetes:helm.sh/v3:Chart                                       redis
 ~   │  ├─ kubernetes:core/v1:ConfigMap                                   redis/redis-scripts                      update      [diff: ~metadata]
 ~   │  ├─ kubernetes:core/v1:ConfigMap                                   redis/redis-configuration                update      [diff: ~metadata]
 ~   │  ├─ kubernetes:core/v1:Secret                                      redis/redis                              update      [diff: ~metadata]
 ~   │  ├─ kubernetes:core/v1:ServiceAccount                              redis/redis                              update      [diff: ~metadata]
 ~   │  ├─ kubernetes:core/v1:ConfigMap                                   redis/redis-health                       update      [diff: ~metadata]
 ~   │  ├─ kubernetes:core/v1:Service                                     redis/redis-metrics                      update      [diff: ~metadata]
 ~   │  ├─ kubernetes:core/v1:Service                                     redis/redis-headless                     update      [diff: ~metadata]
 ~   │  ├─ kubernetes:core/v1:Service                                     redis/redis-master                       update      [diff: ~metadata]
 ~   │  └─ kubernetes:apps/v1:StatefulSet                                 redis/redis-master                       update      [diff: ~metadata,spec]
     ├─ settlemint:prometheus                                             prometheus
     │  ├─ kubernetes:helm.sh/v3:Chart                                    prometheus-promtail
 ~   │  │  ├─ kubernetes:core/v1:ServiceAccount                           monitoring/prometheus-promtail           update      [diff: ~metadata]
 ~   │  │  ├─ kubernetes:core/v1:Secret                                   monitoring/prometheus-promtail           update      [diff: ~metadata]
 ~   │  │  ├─ kubernetes:rbac.authorization.k8s.io/v1:ClusterRole         prometheus-promtail                      update      [diff: ~metadata]
 ~   │  │  ├─ kubernetes:rbac.authorization.k8s.io/v1:ClusterRoleBinding  prometheus-promtail                      update      [diff: ~metadata]
 ~   │  │  └─ kubernetes:apps/v1:DaemonSet                                monitoring/prometheus-promtail           update      [diff: ~metadata,spec]
     │  └─ kubernetes:helm.sh/v3:Chart                                    prometheus-prometheus
 ~   │     ├─ kubernetes:core/v1:PersistentVolumeClaim                    monitoring/prometheus-prometheus-server  update      [diff: ~metadata]
 ~   │     ├─ kubernetes:core/v1:Service                                  monitoring/prometheus-prometheus-server  update      [diff: ~metadata]
 ~   │     ├─ kubernetes:core/v1:ServiceAccount                           monitoring/prometheus-prometheus-server  update      [diff: ~metadata]
 ~   │     ├─ kubernetes:rbac.authorization.k8s.io/v1:ClusterRole         prometheus-prometheus-server             update      [diff: ~metadata]
 ~   │     ├─ kubernetes:core/v1:ConfigMap                                monitoring/prometheus-prometheus-server  update      [diff: ~metadata]
 ~   │     ├─ kubernetes:rbac.authorization.k8s.io/v1:ClusterRoleBinding  prometheus-prometheus-server             update      [diff: ~metadata]
 ~   │     └─ kubernetes:apps/v1:Deployment                               monitoring/prometheus-prometheus-server  update      [diff: ~metadata,spec]
     ├─ kubernetes:helm.sh/v3:Chart                                       rabbitmq
 +-  │  ├─ kubernetes:core/v1:Secret                                      rabbitmq/rabbitmq-config                 replace     [diff: ~data,metadata]
 ~   │  ├─ kubernetes:core/v1:ServiceAccount                              rabbitmq/rabbitmq                        update      [diff: ~metadata]
 ~   │  ├─ kubernetes:rbac.authorization.k8s.io/v1:RoleBinding            rabbitmq/rabbitmq-endpoint-reader        update      [diff: ~metadata]
 ~   │  ├─ kubernetes:core/v1:Service                                     rabbitmq/rabbitmq                        update      [diff: ~metadata]
 ~   │  ├─ kubernetes:core/v1:Service                                     rabbitmq/rabbitmq-headless               update      [diff: ~metadata]
 ~   │  ├─ kubernetes:core/v1:Secret                                      rabbitmq/rabbitmq                        update      [diff: ~metadata]
 ~   │  └─ kubernetes:rbac.authorization.k8s.io/v1:Role                   rabbitmq/rabbitmq-endpoint-reader        update      [diff: ~metadata]
     └─ kubernetes:apps/v1:StatefulSet                                    rabbitmq/rabbitmq                                    1 error


Diagnostics:
  kubernetes:apps/v1:StatefulSet (rabbitmq/rabbitmq):
    error: error reading from server: EOF

  pulumi:pulumi:Stack (bpaas-public-nodes-k8s):
    panic: fatal: An assertion has failed: path: [spec volumeClaimTemplates 0 metadata creationTimestamp]  |  v: <nil>  | old: <nil>  |  oldInput: <nil>
    goroutine 2388 [running]:
    github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.55.0/go/common/util/contract/failfast.go:23
    github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Assertf(0x0?, {0x10480def6?, 0x8?}, {0x1400500dae8?, 0x104d50540?, 0x14005785360?})
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.55.0/go/common/util/contract/assert.go:33 +0xe0
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchValueToDiff(0x1400500f1e8, {0x14005783100, 0x5, 0x8}, {0x0, 0x0}, {0x0, 0x0?}, {0x0?, 0x0?}, ...)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:3178 +0x100
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchMapToDiff(0x14005790ac0?, {0x14005790ac0, 0x4, 0x4}, 0x1400500df80?, 0x5001?, 0x400?, 0x400?, 0x1)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:3289 +0x42c
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchValueToDiff(0x1400500f1e8, {0x14005790ac0, 0x4, 0x4}, {0x104e49760, 0x14005ac1f20}, {0x104e49760, 0x14002d94660?}, {0x104e49760?, 0x14005a829c0?}, ...)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:3201 +0x2c8
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchMapToDiff(0x14005790ac0?, {0x14005790ac0, 0x3, 0x4}, 0x104d50540?, 0x14005ac1590?, 0x1400500f148?, 0x104e49760?, 0x1)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:3280 +0x274
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchValueToDiff(0x1400500f1e8, {0x14005790ac0, 0x3, 0x4}, {0x104e49760, 0x14005ac1ef0}, {0x104e49760, 0x14002d94630?}, {0x104e49760?, 0x14005a82990?}, ...)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:3201 +0x2c8
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchArrayToDiff(0x10480def6?, {0x14005796340, 0x2, 0x2}, {0x14005aca700, 0x1, 0x10e960e40?}, {0x1400568c260, 0x1, 0x5001?}, ...)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:3311 +0x1f0
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchValueToDiff(0x1400500f1e8, {0x14005796340, 0x2, 0x2}, {0x104cf2c20, 0x140051850f8}, {0x104cf2c20, 0x14002f2e7f8?}, {0x104cf2c20?, 0x1400296e600?}, ...)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:3208 +0x228
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchMapToDiff(0x1400578e040?, {0x1400578e040, 0x1, 0x1}, 0x104e02740?, 0x14005785e60?, 0x14005ac73b8?, 0x14005ac7320?, 0x0)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:3280 +0x274
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchValueToDiff(0x1400500f1e8, {0x1400578e040, 0x1, 0x1}, {0x104e49760, 0x14005ac14a0}, {0x104e49760, 0x14005a82ba0?}, {0x104e49760?, 0x14005a79c80?}, ...)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:3201 +0x2c8
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*patchConverter).addPatchMapToDiff(0x14004f5c275?, {0x0, 0x0, 0x0}, 0x0?, 0x0?, 0x0?, 0x14005a750a0?, 0x0)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:3280 +0x274
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.convertPatchToDiff(0x14005ac1410, 0x14005a82ae0, 0x104e49760?, 0x14006013d10?, {0x1400019fc80, 0xc, 0xc})
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:3106 +0xd0
    github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/provider.(*kubeProvider).Diff(0x14000977500, {0x1051e6608, 0x14002a60660}, 0x1400114a310)
    	/home/runner/work/pulumi-kubernetes/pulumi-kubernetes/provider/pkg/provider/provider.go:1597 +0xcb8
    github.com/pulumi/pulumi/sdk/v3/proto/go._ResourceProvider_Diff_Handler.func1({0x1051e6608, 0x14002a60660}, {0x10504df20?, 0x1400114a310})
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.55.0/proto/go/provider_grpc.pb.go:555 +0x74
    github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1({0x1051e6608, 0x140029fcab0}, {0x10504df20, 0x1400114a310}, 0x14002d2fc20, 0x14002b42b28)
    	/home/runner/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x30c
    github.com/pulumi/pulumi/sdk/v3/proto/go._ResourceProvider_Diff_Handler({0x10515b1e0?, 0x14000977500}, {0x1051e6608, 0x140029fcab0}, 0x1400114a2a0, 0x14000539c40)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.55.0/proto/go/provider_grpc.pb.go:557 +0x138
    google.golang.org/grpc.(*Server).processUnaryRPC(0x1400097e780, {0x1051f1aa0, 0x140005824e0}, 0x14001177e60, 0x140001faff0, 0x1067e69c8, 0x0)
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:1340 +0xb7c
    google.golang.org/grpc.(*Server).handleStream(0x1400097e780, {0x1051f1aa0, 0x140005824e0}, 0x14001177e60, 0x0)
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:1713 +0x82c
    google.golang.org/grpc.(*Server).serveStreams.func1.2()
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:965 +0x84
    created by google.golang.org/grpc.(*Server).serveStreams.func1
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.51.0/server.go:963 +0x290
```

### Expected Behavior

It used to work

### Steps to reproduce

This is the stateful set but, timestamp is null, disk is pending

<img width="399" alt="CleanShot 2023-04-12 at 13 50 18@2x" src="https://user-images.githubusercontent.com/16780/231448412-a1a7bebf-25a5-4d21-94b2-1db7bec341bb.png">

### Output of `pulumi about`

```
CLI
Version      3.62.0
Go Version   go1.20.3
Go Compiler  gc

Host
OS       darwin
Version  13.4
Arch     arm64

Backend
Name           pulumi.com
URL            https://app.pulumi.com/roderik
User           roderik
Organizations  roderik, settlemint

Pulumi locates its logs in /var/folders/8k/dr7tcy11347cq606ctpfxmyr0000gn/T/ by default
warning: Failed to read project: no Pulumi.yaml project file found (searching upwards from /Users/roderik/Development/bpaas). If you have not created a project yet, use `pulumi new` to do so: no project file found
warning: Failed to get information about the current stack: no Pulumi.yaml project file found (searching upwards from /Users/roderik/Development/bpaas). If you have not created a project yet, use `pulumi new` to do so: no project file found
```

### Additional context

Cluster is created with https://kurl.sh/ on baremetal Hetzner machines
Using OpenEBS CSI

Weirdly enough the PVC is bound and ready, but still, a bit less strict validation is probably good

![CleanShot 2023-04-12 at 14 02 35@2x](https://user-images.githubusercontent.com/16780/231451169-5518dca2-90e6-40bc-968b-506559899011.png)

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by roderik** [comment]
latest versions of everything

**Comment by rquitales** [comment]
@roderik I'm unable to replicate this issue. I've tried using a minimal StatefulSet manifest, as well as attempting to install the bitnami redis and rabbitmq charts. I can run `pulumi up` and `pulumi refresh` commands successfully without any panics.

Would you be able to provide more information so I can further debug? Would you be able to run `pulumi about` from the directory the Pulumi project itself so we can identify which exact versions you're using as per @kpitzen's suggestion? From the debug log you provided, I don't believe the latest `pulumi-kubernetes` is being used since the latest version of this provider uses v3.63.0 of pulumi/pulumi, whereas the logs you provided indicates an older version of this dependency (v3.55) is being used.

Further, the screenshot you provided of the StatefulSet is as expected, with a null creationTimestamp and pending phase for status, since this is the VolumeClaimTemplate being defined for the PersistentVolume Provisioner to create the PV/PVCs - they will always be empty/meaningless when viewing the StatefulSet object. As you noted, the PV and PVCs are indeed created despite these being their zero value. I've so far identified the panic coming from this line:
https://github.com/pulumi/pulumi-kubernetes/blame/master/provider/pkg/provider/provider.go#L3178 and will need to investigate further to see why this code path is being triggered for the null creationTimestamp template field in your case.

In the meantime, it would be extremely helpful if you could provide a minimal code example so that we can try to replicate this issue, and full outputs of the kubernetes object yamls if possible.

**Comment by rquitales** [comment]
Thanks for the context - this greatly helps my understanding of your setup! To confirm, is this an issue that is no longer occurring for you? Or is it still broken for you, assuming you do not recreate the PVC manually?

**Comment by rquitales** [comment]
Note: downgrading from a P1 since it seems like this issue is isolated to a corner case that so far isn't reproducible. Furthermore, the panic observed is related to inconsistent error handling on type checking, rather than an actual panic.

**Comment by roderik** [comment]
> Thanks for the context - this greatly helps my understanding of your setup! To confirm, is this an issue that is no longer occurring for you? Or is it still broken for you, assuming you do not recreate the PVC manually?

No i went nuclear and deleted the PVC's, afterwards pulumi nicely recreated them with no issues

**Comment by lblackstone** [comment]
I suspect that this is fixed with #2445. We can reopen if this is still failing with the v4 provider.
